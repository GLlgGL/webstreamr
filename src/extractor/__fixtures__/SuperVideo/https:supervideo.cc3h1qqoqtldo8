<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>
    
      Watch Battle Royale (2000)
        
  </title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="Watch video Battle Royale (2000)">
  <meta name="keywords" content="battle, royale, (2000)">

  

  <link rel="stylesheet" href="https://supervideo.cc/assets/css/style.css?v=130">
  <!-- jsdata -->
  <script src="https://supervideo.cc/assets/js/libs.min.js?v=2" type="07528b5ce5c25d81ae05446f-text/javascript"></script>
  <script src="https://supervideo.cc/assets/js/common.js?v=2" type="07528b5ce5c25d81ae05446f-text/javascript"></script>
  <script src="https://supervideo.cc/js/xupload.js?v=4" type="07528b5ce5c25d81ae05446f-text/javascript"></script>
  <!-- favicon -->
  <link rel="apple-touch-icon" href="https://supervideo.cc/assets/images/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="https://supervideo.cc/assets/images/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="https://supervideo.cc/assets/images/favicon/favicon-16x16.png" sizes="16x16">
<!--  <link rel="manifest" href="https://supervideo.cc/assets/images/favicon/manifest.json">-->
  <link rel="icon" href="https://supervideo.cc/assets/images/favicon/favicon.ico">
</head>

<body class="
             
             
             
             
             
               ">
  <header class="header">
    <div class="container-fluid">
      <div class="header__bar flex-start">
        <div class="flex-grow-1">
          <a href="/" class="logo"></a>
        </div>


            <div class="header__right"><a href="https://supervideo.cc/login.html" class="btn  btn_border_default header__btn">
                Login
              </a><a href="https://supervideo.cc/?op=registration" class="btn btn_secondary header__btn">
                Sign Up
              </a></div>
            <div class="header-mobile">
              <button type="button" class="btn-toggle" data-toggle="dropdown">
                <i class="icon icon_user icon_size_18"></i>
              </button>
              <div class="dropdown-menu  dropdown-menu-right header-mobile__dropdown-menu">
                <a href="https://supervideo.cc/login.html" class="btn btn_block btn_border_default">
                  Login
                </a><a href="https://supervideo.cc/?op=registration" class="btn btn_block btn_secondary">
                  Sign Up
                </a>
              </div>
            </div>
        
      </div>
    </div>
  </header>
  
  
  
  <style>
  .modal {
    color: #333;
  }

  .download-codes__copy {
    position: relative;
    margin-bottom: 15px;
    margin-top: 10px;
    float: right;
  }

  .download__player {
    position: relative;
    padding-top: 56.25%;
  }

  .download__info-item {
    margin-bottom: 15px;
    font-size: 14px;
  }

  .download__info-item a {
    color: inherit;
  }

  .download__info-tags {
    margin-bottom: 15px;
  }
  .downloadbox__dropdown-menu li a {
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
  }
  .downloadbox__quality {
    margin-right: 5px;
  }
  .downloadbox__size {font-size: 13px;}
</style>

    
        <!--100% ads-->
		<script src="/9271593.js" type="07528b5ce5c25d81ae05446f-text/javascript"></script>
		<script src="/9271609.js" type="07528b5ce5c25d81ae05446f-text/javascript"></script>
		<script data-cfasync="false" async type="text/javascript" src="//bu.stummelbeelbow.com/r5anaGaI7UQV/115547"></script>
        

        

        

        

        
    

<script language="JavaScript" type="07528b5ce5c25d81ae05446f-text/javascript" CHARSET="UTF-8" src="https://supervideo.cc/js/jquery.cookie.js"></script>
<script type="07528b5ce5c25d81ae05446f-text/javascript">
$.cookie('file_id', '1284502', { expires: 10 });
$.cookie('aff', '11057', { expires: 10 });

</script>

<div class="container-fluid">



<div class="download">
  <div class="download__top" style="background-image: url(https://supervideo.cc/assets/images/bg_download.png);">
    <div class="container">
      

        
          <div id="over_player_msg"  class="alert alert-danger">Video is processing now.<br>Conversion stage: <span id='enc_pp'>...</span></div>
        
        
      

        

        <div class="download__player">
          <div class="overlay cover" style="background-image:url(https://img.supervideo.cc/3h1qqoqtldo8.jpg)">
            
          </div>
        </div>

      <div class="row justify-content-center align-items-center">
        <div class="col-sm-8 mr-auto">
          <h1 class="download__title">
            Battle Royale (2000)
          </h1>
          <ul class="download__info">
            <li><i class="icon icon_clock icon_align_left"></i>
              on
                Jul 24, 2022
            </li>
            <li><i class="icon icon_user icon_align_left"></i><a href="https://supervideo.cc/users/germancloud123">
                germancloud123
              </a></li>
            <li><i class="icon icon_eye icon_align_left"></i>
              373 views
            </li>
            <li><button type="button" class="btn download__flag" data-toggle="modal" data-target="#vid-flag"><i class="icon icon_flag"></i></button>
              <div class="modal fade" id="vid-flag" tabindex="-1" role="dialog" aria-labelledby="modal-title" aria-hidden="true">
                <div class="modal-dialog modal-dialog-centered" role="document">
                  <div class="modal-content">
                    <div class="modal-header">
                      <h5 class="modal-title" id="modal-title">Flag:</h5>
                      <button type="button" class="modal-close" data-dismiss="modal" aria-label="Close">
                        <i class="icon icon_round-close icon_size_20"></i>
                      </button>
                    </div>
                    <div class="modal-body" id="id_flag">
                      <form class="form" method="POST" action="https://supervideo.cc/" onsubmit="if (!window.__cfRLUnblockHandlers) return false; $.post('https://supervideo.cc/',$(this).serialize(),function(code){ eval(code); $('#id_flag').wrapInner('<span></span>').addClass('alert alert-success'); } );return false;" data-cf-modified-07528b5ce5c25d81ae05446f-="">
                      <input type="hidden" name="op" value="report_file">
                      <input type="hidden" name="file_code" value="3h1qqoqtldo8">
                        <div class="form__group">
                          <label class="form__label">
                            Reason
                          </label>
                          <select class="input" name="report_type">
                            <option>
                              Broken video
                            </option>
                            <option>
                              Wrong title/description
                            </option>
                            <option>
                              Copyright restriction
                            </option>
                            <option>
                              Other
                            </option>
                          </select></div>
                        <div class="form__group">
                          <textarea name="report_info" class="input" id="msg" rows="2" placeholder="Details"></textarea>
                        </div>
                          <input type="submit" name="add" value="Send Report" class="btn btn_primary">
                        <button type="submit" class="btn btn_default" data-dismiss="modal">Cancel</button>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </li>
            
          </ul>
        </div>
        <div class="col-sm-auto">
          <ul class="download-social">
            <li><a class="download-social__item download-social__item_facebook" onclick="if (!window.__cfRLUnblockHandlers) return false; window.open(this.href, 'Share Facebook', 'width=640,height=436,toolbar=0,status=0'); return false" href="https://www.facebook.com/sharer/sharer.php?u=https://supervideo.cc/3h1qqoqtldo8" data-cf-modified-07528b5ce5c25d81ae05446f-=""><i class="icon icon_facebook"></i></a></li>
            <li><a class="download-social__item download-social__item_twitter" onclick="if (!window.__cfRLUnblockHandlers) return false; window.open(this.href, 'Share Twitter', 'width=640,height=436,toolbar=0,status=0'); return false" href="https://twitter.com/intent/tweet?text=Battle Royale (2000)" data-cf-modified-07528b5ce5c25d81ae05446f-=""><i class="icon icon_twitter"></i></a></li>
          </ul>
        </div>
        <div class="col-auto">
          <div class="dowonload-rating" id="vote">
            <button type="button" class="btn dowonload-rating__btn dowonload-rating__btn_plus" onclick="if (!window.__cfRLUnblockHandlers) return false; return jah('https://supervideo.cc/?op=vote&file_code=3h1qqoqtldo8&v=up')" title="Like" data-cf-modified-07528b5ce5c25d81ae05446f-=""><i
                class="icon icon_like icon_align_left icon-size_20 dowonload-rating__icon"></i><span id="likes_num">
                0
              </span></button>
            <span class="dowonload-rating__divider"></span>
            <button type="button" class="btn dowonload-rating__btn dowonload-rating__btn_minus" onclick="if (!window.__cfRLUnblockHandlers) return false; return jah('https://supervideo.cc/?op=vote&file_code=3h1qqoqtldo8&v=down')" title="Don&#39;t Like" data-cf-modified-07528b5ce5c25d81ae05446f-=""><i
                class="icon icon_dislike icon_align_left icon-size_20 dowonload-rating__icon"></i><span id="dislikes_num">
                0
              </span></button>
          </div>
        </div>
      </div>
      <div class="row justify-content-center">
        
        
          <div class="col-sm-auto">
            
              <div class="downloadbox">
                <a href="#" data-toggle="dropdown" class="downloadbox__btn"> <i class="icon icon_download downloadbox__icon"></i> <span> <strong class="downloadbox__title">FREE DOWNLOAD</strong> Click to start Download </span> </a>
                <ul class="dropdown-menu downloadbox__dropdown-menu">
                  
                    <li>
                      <a href="#" onclick="if (!window.__cfRLUnblockHandlers) return false; download_video('3h1qqoqtldo8','o','1284502-130-61-1752143560-6f9f5c2cdbce1100170a45ae7f1b6d01')" data-cf-modified-07528b5ce5c25d81ae05446f-=""><i
                          class="icon icon_download icon_size_18 icon_align_left"></i><b class="downloadbox__quality">
                          Original
                        </b><span class="downloadbox__size">
                          1280x696, 1.3 GB
                        </span></a>
                    </li>
                  
                </ul>
              </div>
            
          </div>
        
      </div>
    </div>
  </div>
  <div class="container">
    <div class="row">
      <div class="col-lg-8 order-2 order-lg-1">
        
      </div>
      <div class="col-lg-4 order-1 order-lg-2">
        <div class="dsh-block">
          <ul class="dsh-block__tabs nav">
            <li><a class="active" data-toggle="tab" href="#tab-link">
                Download Link
              </a></li>
            <li><a data-toggle="tab" href="#tab-forum">
                Forum Code
              </a></li>
            <li><a data-toggle="tab" href="#tab-html">HTML</a></li>
            <li><a data-toggle="tab" href="#tab-embed">Embed</a></li>
          </ul>
          <div class="dsh-block__body tab-content clearfix">
            <div class="tab-pane fade show active" id="tab-link">
              <div class="download-codes js-codes">
                <textarea class="input download-codes__area">https://supervideo.cc/3h1qqoqtldo8</textarea>
                <button type="button" class="btn btn_link btn_sm download-codes__copy js-copy-codes"><i class="icon icon_copy icon_align_left"></i><span>Copy</span></button>
              </div>
            </div>
            <div class="tab-pane fade" id="tab-forum">
              <div class="download-codes js-codes">
                <textarea class="input download-codes__area">[URL=https://supervideo.cc/3h1qqoqtldo8][IMG]https://img.supervideo.cc/3h1qqoqtldo8_t.jpg[/IMG]
  Battle Royale (2000)[/URL]
  [1280x696, 01:53:51]</textarea>
                <button type="button" class="btn btn_link btn_sm download-codes__copy js-copy-codes"><i class="icon icon_copy icon_align_left"></i><span>Copy</span></button>
              </div>
            </div>
            <div class="tab-pane fade" id="tab-html">
              <div class="download-codes js-codes">
                <textarea
                  class="input download-codes__area"><a href="https://supervideo.cc/3h1qqoqtldo8"><img src="https://img.supervideo.cc/3h1qqoqtldo8_t.jpg" border=0><br>Battle Royale (2000)</a><br>[1280x696, 01:53:51]</textarea>
                <button type="button" class="btn btn_link btn_sm download-codes__copy js-copy-codes"><i class="icon icon_copy icon_align_left"></i><span>Copy</span></button>
              </div>
            </div>
            <div class="tab-pane fade" id="tab-embed">
              <div class="form__group">
                <div class="row align-items-center text-center justify-content-center">
                  <div class="col-12">
                    <label class="form__label">Embed size: </label>
                  </div>
                  <div class="col-auto">
                    <div class="row no-gutters align-items-center">
                      <div class="col-auto">
                        <input type="text" class="input" id="iew" value="640" size="3">
                      </div>
                      <div class="col-auto text-muted"><i class="icon icon_close icon_size_8 icon_align_left icon_align_right"></i></div>
                      <div class="col-auto"><input class="input" type="text" id="ieh" value="360" size="3"></div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="download-codes js-codes">
                <textarea id="iet" class="input download-codes__area"><div style="position:relative;padding-bottom:56%;padding-top:20px;height:0;"><irame src="https://supervideo.cc/e/3h1qqoqtldo8" frameborder=0 marginwidth=0 marginheight=0 scrolling=no width=640 height=360 allowfullscreen style="position:absolute;top:0;left:0;width:100%;height:100%;"></iframe></div></textarea>
                <button type="button" class="btn btn_link btn_sm download-codes__copy js-copy-codes"><i class="icon icon_copy icon_align_left"></i><span>Copy</span></button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
</div>


  <script type="07528b5ce5c25d81ae05446f-text/javascript">
  
  encStatus('1284502',1);
  
</script>

    
    <footer class="footer">
      <div class="container-fluid">
        <div class="footer__copyright">© 2017 - 2025 SuperVideo | Fast HLS Video Streaming Service
        </div>
        <ul class="footer__menu">
          <li><a href="https://supervideo.cc">
              Home
            </a></li>
          
          
          <li><a href="https://supervideo.cc/faq">
              FAQ
            </a></li>
          <li><a href="https://supervideo.cc/tos">
              Terms of service
            </a></li>
          
            
              <li><a href="https://supervideo.cc/premium">
                  Trust Uploader
                </a></li>
            
          
          <li><a href="https://supervideo.cc/make_money.html">
              Make Money
            </a></li>
          <li><a href="https://supervideo.cc/checkfiles.html">
              Link Checker
            </a></li>
          <li><a href="https://supervideo.cc/contacts">
              Contact Us
            </a></li>
        </ul>
        <div class="text-center" style="padding-top: 8px;">
        <!--  <a href="http://www.xvstheme.com/" rel="nofollow" title="Designed by XVSTheme" class="small text-reset" style="display:inline-flex; align-items: center">
            <span style="margin-right: 4px;">Designed by</span>
            <svg xmlns="http://www.w3.org/2000/svg" width="78.1" height="11" viewBox="0 0 78.1 11" xml:space="preserve"><path d="M73.5 7.5h-5.7c0 1.2.7 1.9 1.8 1.9.8 0 1.4-.4 1.6-1h2.4c-.6 1.7-1.9 2.6-3.9 2.6-2.8 0-4.2-1.4-4.2-4.1 0-2.8 1.3-4.1 4.1-4.1 2.7 0 4 1.3 4 3.8 0 .3 0 .6-.1.9zm-3.9-3.1c-1.1 0-1.6.4-1.7 1.6h3.4c-.2-1.1-.7-1.6-1.7-1.6zM62 6.3c0-1-.4-1.4-1.3-1.4-1 0-1.5.5-1.5 1.6v4.2h-2.3V6.3c0-1-.4-1.4-1.3-1.4-1 0-1.5.5-1.5 1.6v4.2h-2.3V3.1l2.1-.1.2 1c.4-.7 1.2-1.1 2.2-1.1 1.3 0 2.2.5 2.6 1.4.5-.9 1.3-1.4 2.5-1.4 2 0 3 1.1 3 3.4v4.5H62V6.3zM44.6 7.5c0 1.2.7 1.9 1.8 1.9.8 0 1.4-.4 1.6-1h2.4c-.5 1.7-1.8 2.6-3.8 2.6-2.8 0-4.2-1.4-4.2-4.1 0-2.8 1.3-4.1 4.1-4.1 2.7 0 4 1.3 4 3.8 0 .3 0 .6-.1.9h-5.8zm1.8-3.1c-1.1 0-1.6.4-1.7 1.6H48c-.1-1.1-.6-1.6-1.6-1.6zm-7.6 1.9c0-1-.4-1.4-1.4-1.4-1.2 0-1.6.5-1.6 1.6v4.2h-2.3V0h2.3v3.8c.4-.6 1.2-1 2.2-1 2.1 0 3.1 1.1 3.1 3.4v4.5h-2.4c.1.1.1-4.4.1-4.4zM27.4 7.6V5.2h-1.1V3.5l1.1-.4V1.4h2.3v1.7H32v2.1h-2.3v2.3c0 .9.3 1.2 1.2 1.2h1v2.1h-1.3c-2.3 0-3.2-.9-3.2-3.2zm-4.6-1.5c2.2.4 2.8.9 2.8 2.4S24.1 11 22 11c-2.4 0-3.8-.9-4.1-2.6h2.3c.2.6.7 1 1.7 1 .8 0 1.3-.4 1.3-.8 0-.5-.2-.7-1.2-.8l-1.5-.2c-1.8-.3-2.6-1-2.6-2.1C17.9 4 19.3 3 21.5 3c2.3 0 3.7.9 3.8 2.4h-2.4c-.1-.5-.7-.8-1.6-.8-.7 0-1.2.3-1.2.8 0 .4.2.6 1 .7.3-.1 1.7 0 1.7 0zM13.2 11c-2.2 0-3.1-1-3.5-3.5L9 3.1h2.3l.7 4.3c.2 1.1.5 1.5 1.2 1.5s1-.4 1.2-1.5l.6-4.3h2.4l-.7 4.4c-.4 2.5-1.4 3.5-3.5 3.5zm-7.3-.2L4.4 8.4l-1.6 2.3H0l2.8-3.9L.1 3.1h2.8l1.4 2.3 1.4-2.3h2.8L6 6.9l2.8 3.9H5.9z" fill="#000222"/><path d="M78.1 8.7h-2.6V11h2.6V8.7z" fill="#28e98c"/></svg>
          </a> -->
        </div>
      </div>
    </footer>
<script src="/cdn-cgi/scripts/7d0fa10a/cloudflare-static/rocket-loader.min.js" data-cf-settings="07528b5ce5c25d81ae05446f-|49" defer></script></body>
</html>