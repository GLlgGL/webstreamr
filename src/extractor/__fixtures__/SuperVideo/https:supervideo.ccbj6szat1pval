<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>
    
      Watch the walking dead dead city s02e05 german 1080p web x264 wvf
        
  </title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="Watch video the walking dead dead city s02e05 german 1080p web x264 wvf">
  <meta name="keywords" content="the, walking, dead, dead, city, s02e05, german, 1080p, web, x264, wvf">

  

  <link rel="stylesheet" href="https://supervideo.cc/assets/css/style.css?v=130">
  <!-- jsdata -->
  <script src="https://supervideo.cc/assets/js/libs.min.js?v=2" type="74b787644b21e08d1cd03347-text/javascript"></script>
  <script src="https://supervideo.cc/assets/js/common.js?v=2" type="74b787644b21e08d1cd03347-text/javascript"></script>
  <script src="https://supervideo.cc/js/xupload.js?v=4" type="74b787644b21e08d1cd03347-text/javascript"></script>
  <!-- favicon -->
  <link rel="apple-touch-icon" href="https://supervideo.cc/assets/images/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="https://supervideo.cc/assets/images/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="https://supervideo.cc/assets/images/favicon/favicon-16x16.png" sizes="16x16">
<!--  <link rel="manifest" href="https://supervideo.cc/assets/images/favicon/manifest.json">-->
  <link rel="icon" href="https://supervideo.cc/assets/images/favicon/favicon.ico">
</head>

<body class="
             
             
             
             
             
               ">
  <header class="header">
    <div class="container-fluid">
      <div class="header__bar flex-start">
        <div class="flex-grow-1">
          <a href="/" class="logo"></a>
        </div>


            <div class="header__right"><a href="https://supervideo.cc/login.html" class="btn  btn_border_default header__btn">
                Login
              </a><a href="https://supervideo.cc/?op=registration" class="btn btn_secondary header__btn">
                Sign Up
              </a></div>
            <div class="header-mobile">
              <button type="button" class="btn-toggle" data-toggle="dropdown">
                <i class="icon icon_user icon_size_18"></i>
              </button>
              <div class="dropdown-menu  dropdown-menu-right header-mobile__dropdown-menu">
                <a href="https://supervideo.cc/login.html" class="btn btn_block btn_border_default">
                  Login
                </a><a href="https://supervideo.cc/?op=registration" class="btn btn_block btn_secondary">
                  Sign Up
                </a>
              </div>
            </div>
        
      </div>
    </div>
  </header>
  
  
  
  <script type="74b787644b21e08d1cd03347-text/javascript">
  document.body.classList.remove('is_home')
</script>

 
<div class="container">


	<div class="alert alert-warning">
		This video can be watched as embed only
	</div>

</div>


 

    
    <footer class="footer">
      <div class="container-fluid">
        <div class="footer__copyright">Â© 2017 - 2025 SuperVideo | Fast HLS Video Streaming Service
        </div>
        <ul class="footer__menu">
          <li><a href="https://supervideo.cc">
              Home
            </a></li>
          
          
          <li><a href="https://supervideo.cc/faq">
              FAQ
            </a></li>
          <li><a href="https://supervideo.cc/tos">
              Terms of service
            </a></li>
         <!--- 
            
              <li><a href="https://supervideo.cc/premium">
                  Trust Uploader
                </a></li>
            
           --->
          <li><a href="https://supervideo.cc/make_money.html">
              Make Money
            </a></li>
       <!---   <li><a href="https://supervideo.cc/checkfiles.html">
              Link Checker
            </a></li> --->
          <li><a href="https://supervideo.cc/contacts">
              Contact Us
            </a></li>
        </ul>
        <div class="text-center" style="padding-top: 8px;">
        </div>
      </div>
    </footer>
<script src="/cdn-cgi/scripts/7d0fa10a/cloudflare-static/rocket-loader.min.js" data-cf-settings="74b787644b21e08d1cd03347-|49" defer></script><script>(function(){function c(){var b=a.contentDocument||a.contentWindow.document;if(b){var d=b.createElement('script');d.innerHTML="window.__CF$cv$params={r:'9765f2e01875d376',t:'MTc1NjQwNjU0OS4wMDAwMDA='};var a=document.createElement('script');a.nonce='';a.src='/cdn-cgi/challenge-platform/scripts/jsd/main.js';document.getElementsByTagName('head')[0].appendChild(a);";b.getElementsByTagName('head')[0].appendChild(d)}}if(document.body){var a=document.createElement('iframe');a.height=1;a.width=1;a.style.position='absolute';a.style.top=0;a.style.left=0;a.style.border='none';a.style.visibility='hidden';document.body.appendChild(a);if('loading'!==document.readyState)c();else if(window.addEventListener)document.addEventListener('DOMContentLoaded',c);else{var e=document.onreadystatechange||function(){};document.onreadystatechange=function(b){e(b);'loading'!==document.readyState&&(document.onreadystatechange=e,c())}}}})();</script></body>
</html>