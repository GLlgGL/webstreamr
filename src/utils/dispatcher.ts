import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
