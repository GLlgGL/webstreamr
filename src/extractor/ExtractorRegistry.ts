import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
