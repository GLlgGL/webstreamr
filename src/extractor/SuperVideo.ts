import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { NotFoundError } from '../error';
import { Context, Format, Meta, UrlResult } from '../types';
import { extractUrlFromPacked, guessHeightFromPlaylist } from '../utils';
import { Extractor } from './Extractor';

export class SuperVideo extends Extractor {
  public readonly id = 'supervideo';

  public readonly label = 'SuperVideo';

  public override readonly ttl: number = 10800000; // 3h

  public supports(_ctx: Context, url: URL): boolean {
    return null !== url.host.match(/supervideo/);
  }

  public override normalize(url: URL): URL {
    return new URL(url.href.replace('/e/', '/').replace('/k/', '/').replace('/embed-', '/'));
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { Referer: meta.referer ?? url.origin };

    const html = await this.fetcher.text(ctx, url, { headers });

    if (html.includes('This video can be watched as embed only')) {
      return await this.extractInternal(ctx, new URL(`/e${url.pathname}`, url.origin), meta);
    }

    if (/'The file was deleted|The file expired|Video is processing/.test(html)) {
      throw new NotFoundError();
    }

    const m3u8Url = extractUrlFromPacked(html, [/sources:\[{file:"(.*?)"/]);

    const heightAndSizeMatch = html.match(/\d{3,}x(\d{3,}), ([\d.]+ ?[GM]B)/);
    const size = heightAndSizeMatch ? bytes.parse(heightAndSizeMatch[2] as string) as number : undefined;
    const height = heightAndSizeMatch
      ? parseInt(heightAndSizeMatch[1] as string)
      : await guessHeightFromPlaylist(ctx, this.fetcher, m3u8Url, { headers });

    const $ = cheerio.load(html);
    const title = $('.download__title').text().trim();

    return [
      {
        url: m3u8Url,
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          title,
          ...(size && { bytes: size }),
          ...(height && { height }),
        },
      },
    ];
  };
}
