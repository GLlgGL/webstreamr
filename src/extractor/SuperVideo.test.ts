import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { SuperVideo } from './SuperVideo';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new SuperVideo(new FetcherMock(`${__dirname}/__fixtures__/SuperVideo`))]);

const ctx = createTestContext();

describe('SuperVideo', () => {
  test('supervideo.cc /e/', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://supervideo.cc/e/q7i0sw1oytw3'))).toMatchSnapshot();
  });

  test('supervideo.tv /embed-/', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://supervideo.tv/embed-1p0m1fi9mok8.html'))).toMatchSnapshot();
  });

  test('embed only', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://supervideo.cc/e/bj6szat1pval'))).toMatchSnapshot();
  });

  test('deleted or expired file', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://supervideo.cc/ndf5shmy9lpt'))).toMatchSnapshot();
  });

  test('processing video', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://supervideo.cc/3h1qqoqtldo8'))).toMatchSnapshot();
  });
});
