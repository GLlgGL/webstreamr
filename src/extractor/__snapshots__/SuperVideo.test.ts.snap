// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SuperVideo deleted or expired file 1`] = `[]`;

exports[`SuperVideo embed only 1`] = `
[
  {
    "format": "hls",
    "label": "SuperVideo",
    "meta": {
      "countryCodes": [],
      "height": 1080,
      "title": "",
    },
    "sourceId": "supervideo_",
    "ttl": 10800000,
    "url": "https://hfs308.serversicuro.cc/hls/,dnzpe7rf5xg4a3gyva3h5orlyw23xosfzwunl5xbmjpz2l6t3doh5zbwynna,.urlset/master.m3u8",
  },
]
`;

exports[`SuperVideo processing video 1`] = `[]`;

exports[`SuperVideo supervideo.cc /e/ 1`] = `
[
  {
    "format": "hls",
    "label": "SuperVideo",
    "meta": {
      "bytes": 1073741824,
      "countryCodes": [],
      "height": 720,
      "title": "des-teufels-bad-2024",
    },
    "sourceId": "supervideo_",
    "ttl": 10800000,
    "url": "https://hfs309.serversicuro.cc/hls/,dnzpfi3d27g4a3gyvbmh5klwtl65qh654hyjtgupd6p577fze64cp3rsmdyq,.urlset/master.m3u8",
  },
]
`;

exports[`SuperVideo supervideo.tv /embed-/ 1`] = `
[
  {
    "format": "hls",
    "label": "SuperVideo",
    "meta": {
      "bytes": 219571814,
      "countryCodes": [],
      "height": 344,
      "title": "Babylon 5 2x3",
    },
    "sourceId": "supervideo_",
    "ttl": 10800000,
    "url": "https://hfs307.serversicuro.cc/hls/,dnzpaalv5xg4a3gyvavh52jryoft7o5cd5nm7cwtnu74kmc4ohiklh5qwq4q,.urlset/master.m3u8",
  },
]
`;
